src/opam.nix

```diff
@@ -562,6 +562,19 @@ in rec {
 
   # remove non-dune dependant packages
   opamRepositoryFiltered =
+    # TODO is there a way to do this outside a derivation?
```

Comment:
No, there's not. IFD is the only way.

Reply:
(it might be a bit faster to do all the `opam2json`-ning in one derivation and then read the combined output, though).

Reply:
Yeah, I was thinking the same thing. I'll give it a shot.

Reply:
If this works, I'm wondering if materialization might be a way of speeding up deployments...

Reply:
An aside: it would be nice if we could use just the version solving functionality of opam without pulling the whole program in. I might look into the feasibility of a refactor of opam to support this.

Reply:
Okay, I've hacked something together that does this:
```
  # remove non-dune dependant packages
  opamRepositoryFiltered =
    let
      keepPackages = [ "ocaml" "dune" "ocaml-solo5" "ocaml-src" "solo5"
      "ocaml-config" "ocaml-system" "ocaml-variants" "base-bytes" ];
    in bootstrapPackages.stdenv.mkDerivation {
      name = "opamRepositoryFiltered";
      src = opamRepository;
      buildInputs = with bootstrapPackages; [ jq ];
      installPhase = ''
        # look at package version directories
        # (e.g. /nix/store/<hash>-opam-repository-modulo-overlay/packages/<name>/<name>.<version>)
        for f in $(find packages -mindepth 2 -maxdepth 2); do
          echo $f
          # special cases as we will filter them before creating monorepo
          # just needed so `opam admin list` doesn't complain
          # TODO parse queries opam files to a list of dependancies with 
          # build=false so only pick up monorepo deps and we remove this workarounds
          # TODO or just use opam monorepo solver...
          if $(echo "$f" | awk 'BEGIN {
              split("${concatStringsSep " " keepPackages}", parts);
              for (i in parts) dict[parts[i]]=""
            } { exit !($1 in dict) }'); then
            # do nothing
            echo
          # if package depends on dune
          # TODO properly recursively parse vals of depends
          elif $(${opam2json}/bin/opam2json "$f/opam" |\
            jq '[ .depends | flatten | .[] | if type=="string" then . else .val end ] |
              if index("dune") == null then halt_error(1) else halt end'); then
              mkdir -p "$out/$f"
              cp -r "$f" "$out/$f"
          fi
        done
      '';
    };
 ```
 But it's also *very* slow (waiting on exact timings now. I think using `opam monorepo`'s solver would be a better solution.

Reply:
Okay, this creates a very IO bound process that takes ~half an hour to complete 😄 I'm going to look at using `opam monorepo lock` and parsing the lockfile to solver versions.

Reply:
That sounds reasonable.